File: leetcode/84.LargestRectangleInHistogram.js
/**
 * @param {number[]} heights
 * @return {number}
 */
const largestRectangleArea = function (heights) {
  let max = 0
  const stack = []

  for (let i = 0; i < heights.length; i++) {
    const height = heights[i]
    let start = i

    while (stack.length && stack[stack.length - 1][1] > height) {
      const [index, h] = stack.pop()

      max = Math.max(max, h * (i - index))
      start = index
    }
    stack.push([start, height])
  }

  for (const [index, height] of stack) {
    max = Math.max(max, height * (heights.length - index))
  }

  return max
}

console.log(largestRectangleArea([2, 1, 5, 6, 2, 3]))
// console.log(largestRectangleArea([2,4]))

File: leetcode/84.LargestRectangleInHistogram.go
package leetcode

import "math"

func largestRectangleArea(heights []int) int {
	max := 0
	var stack [][]int

	for i := 0; i < len(heights); i++ {
		height := heights[i]
		start := i

		for len(stack) > 0 && stack[len(stack)-1][1] > height {
			popped := stack[len(stack)-1]
			stack = stack[:len(stack)-1]

			index, h := popped[0], popped[1]

			max = int(math.Max(float64(max), float64(h*(i-index))))
			start = index
		}
		stack = append(stack, []int{start, height})
	}

	for _, s := range stack {
		index, height := s[0], s[1]
		max = int(math.Max(float64(max), float64(height*(len(heights)-index))))
	}

	return max
}
